viewer-prototype/src/browser/trace-viewer/components/utils/filter-tree-component.tsx

```diff
@@ -0,0 +1,230 @@
+        const toggledNode = this.getNode(nodes[0], node.id);
+        toggledNode.expanded = !node.expanded;
+        this.setState({treeNodes: nodes})
+    }
+
+    getAllChildrenIds = (node: TreeNode, ids: number[]) => {
+        if (node.type === 'leaf') {
+            return ids.push(node.id);
+        } else if (node.children.length) {
+            ids.push(node.id);
+            node.children.forEach((child: TreeNode) => {
+                this.getAllChildrenIds(child, ids);
+            })
+        }
+        return ids;
+    }
+
+    areAllSiblingsChecked = (node: TreeNode) => {
+        const currentCheckedIds = this.state.checkedIDs.concat([node.id]);
+        const parentNode = this.getNode(this.state.treeNodes[0], node.parentId);
+        return parentNode.children.every((child: TreeNode) => {
+            return currentCheckedIds.includes(child.id);
+        });
+    }
+
+    handleCheck = (id: number) => {
+        let checkedIds: number[] = [];
+        const checkedNode = this.getNode(this.state.treeNodes[0], id);
+        if (checkedNode.type === 'node' && checkedNode.children.length) {
+            checkedIds = checkedIds.concat(this.getAllChildrenIds(checkedNode, []));
+        } else {
+            checkedIds.push(id);
+            if (this.areAllSiblingsChecked(checkedNode)) {
+                checkedIds.push(checkedNode.parentId);
+            }
+        }
+        this.props.onChecked(checkedIds);
+    }
+
+    render() {
+        if (!this.props.nodes || !this.props.checkedSeries) return null;
+        const rootNodes = this.getRootNodes();
+        if (rootNodes !== undefined && rootNodes.length > 0) {
+            return (
+                <div>
+                    {rootNodes.map((node, i) => {
+                        return (
+                            <TreeNodeComponent 
+                                node={node}
+                                key={'node-'+node.id}
+                                getChildNodes={this.getChildNodes}
+                                level={0}
+                                onToggle={this.onToggle}
+                                checkedSeries={this.props.checkedSeries}
+                                onChecked={this.handleCheck}
+                            />
+                        )
+                        
+                    })}
+                </div>
+            )
+        }
+        return (<div></div>)
+    }
+}
+
+type TreeNodeComponentProps = {
+    node: TreeNode;
+    level: number;
+    checkedSeries: number[];
+    getChildNodes: (node: TreeNode) => TreeNode[];
+    onToggle: (node: TreeNode) => void;
+    onChecked: (id: number) => void;
+}
+
+class TreeNodeComponent extends React.Component<TreeNodeComponentProps> {
+    constructor(props: TreeNodeComponentProps) {
+        super(props);
+    }
+
+    getPaddingLeft = (level: number, type: string) => {
+        let paddingLeft = level * 15;
+        if (type === 'leaf') paddingLeft += 15;
```

Comment:
type should be either an enum or extracted to some constants. Also is it necessary or does the child count give the answer if it's a leaf or a node?